src/room007/models/word_tag_predictor.py

```diff
@@ -1,100 +1,130 @@
 # vim: set fileencoding=utf-8 :
 
-import itertools
-from sklearn.feature_extraction.text import TfidfVectorizer
-from sklearn import linear_model
-
-import nltk
+from copy import deepcopy
+from itertools import chain
+import logging
 import string
 
-from sklearn.neighbors import KNeighborsClassifier
-from sklearn.tree import DecisionTreeClassifier
-from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
-from sklearn.naive_bayes import GaussianNB
+logger = logging.getLogger(__name__)
+
+import nltk
+from sklearn import linear_model
 from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
+from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
+from sklearn.feature_extraction.text import TfidfVectorizer
 from sklearn.linear_model import LogisticRegression
+from sklearn.naive_bayes import GaussianNB
+from sklearn.neighbors import KNeighborsClassifier
+from sklearn.tree import DecisionTreeClassifier
 
 from room007.models import model
+from room007.util import time_function
+
 
+stop_words = (set(nltk.corpus.stopwords.words('english'))
+              .union(string.printable))
 
-stop_words = nltk.corpus.stopwords.words('english') + [x for x in string.printable]
+
+def identity(x):
+    return x
+
+
+def tokenize(sentence):
+    return [token for token in sentence.split() if token not in stop_words]
 
 
 class Features(object):
     def __init__(self, functional_test=False, changes=False):
         self.functional_test = functional_test
         self.tf_idf_vectorizer = None
         self.changes = changes
+        self._changes_int = 1 if changes else 0
+        self._fit_data_id = None
 
     def fit(self, train_data):
-        self._train_tf_idf_vectorizer(train_data)
+        if self._fit_data_id != id(train_data):
+            data_cp = deepcopy(train_data)
```

Comment:
Replying to the first comment: yes, it will always be true in the current usage scenario. However, the protocol does not forbid calling `Features.fit` repeatedly, in which case expanding the data frame for the derived columns will need to be done again in any subsequent call, unless it's called on the very same data frame as earlier. For this, I decided to assume that when fitting `Features` to the very same object as they have been fitted to earlier, the result of fitting would be the same, which neglects the possibility of modifying the data frame in place.